update/containers.go

```diff
@@ -0,0 +1,171 @@
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/image"
+	"github.com/weaveworks/flux/resource"
+)
+
+// ContainerSpecs defines the spec for a `containers` manifest update.
+type ContainerSpecs struct {
+	Kind             ReleaseKind
+	ContainerSpecs   map[flux.ResourceID][]ContainerUpdate
+	IgnoreMismatches bool
+}
+
+// CalculateRelease computes required controller updates to satisfy this specification.
+// It returns an error if any spec calculation fails unless `IgnoreMismatches` is true.
+func (s ContainerSpecs) CalculateRelease(rc ReleaseContext, logger log.Logger) ([]*ControllerUpdate, Result, error) {
+	all, results, err := s.selectServices(rc)
+	if err != nil {
+		return nil, results, err
+	}
+	updates := s.controllerUpdates(results, all)
+
+	failures := 0
+	successes := 0
+	for _, res := range results {
+		switch res.Status {
+		case ReleaseStatusFailed:
+			failures++
+		case ReleaseStatusSuccess:
+			successes++
+		}
+	}
+	if failures > 0 {
+		return updates, results, errors.New("cannot satisfy specs")
+	}
+	if successes == 0 {
+		return updates, results, errors.New("no changes found")
+	}
+
+	return updates, results, nil
+}
+
+func (s ContainerSpecs) selectServices(rc ReleaseContext) ([]*ControllerUpdate, Result, error) {
+	results := Result{}
+	var rids []flux.ResourceID
+	for rid := range s.ContainerSpecs {
+		rids = append(rids, rid)
+	}
+	all, err := rc.SelectServices(results, []ControllerFilter{&IncludeFilter{IDs: rids}}, nil)
+	if err != nil {
+		return nil, results, err
+	}
+	return all, results, nil
+}
+
+func (s ContainerSpecs) controllerUpdates(results Result, all []*ControllerUpdate) []*ControllerUpdate {
+	var updates []*ControllerUpdate
+	for _, u := range all {
+		cs, err := u.Controller.ContainersOrError()
+		if err != nil {
+			results[u.ResourceID] = ControllerResult{
+				Status: ReleaseStatusFailed,
+				Error:  err.Error(),
+			}
+			continue
+		}
+
+		containers := map[string]resource.Container{}
+		for _, spec := range cs {
+			containers[spec.Name] = spec
+		}
+
+		var mismatch, notfound []string
+		var containerUpdates []ContainerUpdate
+		for _, spec := range s.ContainerSpecs[u.ResourceID] {
+			container, ok := containers[spec.Container]
+			if !ok {
+				notfound = append(notfound, spec.Container)
+				continue
+			}
+
+			// An empty spec for the current image skips the precondition
+			empty := image.Ref{}
+			if spec.Current != empty && container.Image != spec.Current {
+				mismatch = append(mismatch, spec.Container)
+				continue
+			}
+
+			if container.Image == spec.Target {
+				// Nothing to update
+				continue
+			}
+
+			containerUpdates = append(containerUpdates, spec)
+		}
+
+		skippedError := ImageUpToDate
+		notFoundError := fmt.Sprintf(ContainerNotFound, strings.Join(notfound, ", "))
+		mismatchError := fmt.Sprintf(ContainerTagMismatch, strings.Join(mismatch, ", "))
+
+		if s.IgnoreMismatches {
```

Comment:
The only place this matters if in the case `len(containerUpdates) == 0` below, so you could move it there.